internal/controller/management/serviceaccount/serviceaccounts.go

```diff
@@ -0,0 +1,229 @@
+package upgrade
+
+import (
+	"context"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	kubeerr "k8s.io/apimachinery/pkg/api/errors"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	kargoapi "github.com/akuity/kargo/api/v1alpha1"
+	"github.com/akuity/kargo/internal/logging"
+)
+
+const (
+	ComponentLabelKey = "app.kubernetes.io/component"
+	ControllerLabelValue = "controller"
+)
+
+type ReconcilerConfig struct {
+	KargoNamespace string `envconfig:"KARGO_NAMESPACE" required:"true"`
+}
+
+// ServiceAccountReconciler Reconciles for ServiceAccounts
+type ServiceAccountReconciler struct {
+	cfg ReconcilerConfig
+	client.Client
+}
+
+// SetupReconcilerWithManager initializes a reconciler
+// and registers that reconciler with the provided Manager.
+func SetupReconcilerWithManager(kargoMgr manager.Manager) error {
+	return ctrl.NewControllerManagedBy(kargoMgr).
+		For(&corev1.ServiceAccount{}).
+		WithEventFilter(predicate.Funcs{
+			CreateFunc: func(e event.CreateEvent) bool {
+				return hasControllerLabel(e.Object.(*corev1.ServiceAccount))
+			},
+			DeleteFunc: func(e event.DeleteEvent) bool {
+				return hasControllerLabel(e.Object.(*corev1.ServiceAccount))
+			},
+			UpdateFunc: func(e event.UpdateEvent) bool {
+				oldSA := e.ObjectOld.(*corev1.ServiceAccount)
+				newSA := e.ObjectNew.(*corev1.ServiceAccount)
+				return hasControllerLabel(oldSA) != hasControllerLabel(newSA)
+			},
+			GenericFunc: func(event.GenericEvent) bool {
+				return false
+			},
+		}).
+		Complete(&ServiceAccountReconciler{
+			Client:         kargoMgr.GetClient(),
+		})
```

Comment:
I recommend creating a contructor-like function like `newReconciler` and using it here. For non-trivial components where there is high potential for mis-configuration, the signature of such a function helps ensure that the calling code has indeed injected anything required and the logic of such a function provides a better/safer guarantee of proper initialization.

Reply:
It's never _strictly_ needed. It's just a good idea when there's a lot involved in initializing something properly. Right now it may seem there _isn't_ a lot involved in that... but I know how things tend to go with reconcilers over time.